quote/gap.go

```diff
@@ -0,0 +1,147 @@
+	"strings"
+	"time"
+
+	"github.com/go-gota/gota/dataframe"
+	lib "github.com/wopta/goworkspace/lib"
+	"github.com/wopta/goworkspace/models"
+)
+
+func GapFx(w http.ResponseWriter, r *http.Request) (string, interface{}, error) {
+	req := lib.ErrorByte(io.ReadAll(r.Body))
+	var data models.Policy
+	defer r.Body.Close()
+	e := json.Unmarshal(req, &data)
+
+	authToken, err := models.GetAuthTokenFromIdToken(r.Header.Get("Authorization"))
+	lib.CheckError(err)
+	res, e := Gap(authToken.Role, data)
+	s, e := json.Marshal(res)
+	return string(s), nil, e
+
+}
+
+func Gap(role string, data models.Policy) (models.Policy, error) {
+	var err error
+
+	//sellable rules need to be called here
+
+	baseMatrix := lib.GetFilesByEnv("quote/gap_matrix_base.csv")
+	completeMatrix := lib.GetFilesByEnv("quote/gap_matrix_complete.csv")
+	provincesMatrix := lib.GetFilesByEnv("enrich/provinces.csv")
+
+	baseMatrixDF := lib.CsvToDataframe(baseMatrix)
+	completeMatrixDF := lib.CsvToDataframe(completeMatrix)
+	provincesMatrixDF := lib.CsvToDataframe(provincesMatrix)
+
+	//get the:
+	// - vehicle owner residence
+	// - vechicle value
+	var residenceCode string = data.Assets[len(data.Assets)-1].Person.Residence.Locality
+	var vehicleValue int64 = data.Assets[len(data.Assets)-1].Vehicle.PriceValue
+
+	//get the residence area
+	residenceArea := getResidentArea(provincesMatrixDF, residenceCode)
+
+	//get the duration
+	duration := getDuration(data.StartDate, data.EndDate)
+
+	fmt.Printf("valore base e' %d\n", duration)
+
+	//get the base and complete multipliers
+	baseGapMultiplierFloat, completeGapMultiplierFloat := getGapMultipliers(baseMatrixDF, completeMatrixDF, duration, residenceArea)
+
+	//get the matrix area row
+	fmt.Printf("valore base e' %f\n", baseGapMultiplierFloat)
+	fmt.Printf("valore completo e' %f\n", completeGapMultiplierFloat)
+	fmt.Printf("valore base veicolo*tax e' %f\n", baseGapMultiplierFloat*float64(vehicleValue))
+	fmt.Printf("valore completo veicolo*tax e' %f\n", completeGapMultiplierFloat*float64(vehicleValue))
+
+	//set the offer in the policy and round to 2 decimal number
+
+	data = setOffersPrices(data, vehicleValue, baseGapMultiplierFloat, completeGapMultiplierFloat)
+
+	roundGapOffersPrices(data.OffersPrices)
+
+	return data, err
+}
+
+func roundGapOffersPrices(offersPrices map[string]map[string]*models.Price) {
+	for offerKey, offerValue := range offersPrices {
+		for paymentKey := range offerValue {
+			offersPrices[offerKey][paymentKey].Net = lib.RoundFloat(offersPrices[offerKey][paymentKey].Net, 2)
+		}
+	}
+}
+
+func setOffersPrices(data models.Policy, vehicleValue int64, baseGapMultiplierFloat float64, completeGapMultiplierFloat float64) models.Policy {
+	data.OffersPrices = make(map[string]map[string]*models.Price)
+
+	data.OffersPrices["base"] = map[string]*models.Price{
+		"singleInstallment": {
+			Net:      baseGapMultiplierFloat * float64(vehicleValue),
+			Tax:      0.0,
+			Gross:    0.0,
+			Delta:    0.0,
+			Discount: 0.0,
+		},
```

Comment:
we could extract the construction of an offer price into a function (the only difference is the multiplier) that internally do all the math for calculating Net, Tax and Gross values